metals/src/main/scala/scala/meta/internal/parsing/JavaTokens.scala

```diff
@@ -0,0 +1,159 @@
+package scala.meta.internal.parsing
+
+import scala.meta.inputs.Input
+import scala.util.Try
+import org.eclipse.jdt.core.ToolFactory
+import org.eclipse.jdt.core.compiler.ITerminalSymbols
+import scala.meta.tokens.Token
+import scala.meta.dialects
+import org.eclipse.jdt.core.compiler.IScanner
+import scala.meta.tokens.Tokens
+import scala.collection.mutable.ArrayBuffer
+import scala.meta.Dialect
+
+object JavaTokens {
+
+  /**
+   * Best effort translation of java tokens to the scalameta tokens, should work for most purposes.
+   *
+   * @param source to tokenize
+   * @return tokenized source
+   */
+  def tokenize(source: Input): Try[Tokens] = Try {
+
+    val scanner = ToolFactory.createScanner(
+      /*tokenizeComments = */ true, /*tokenizeWhiteSpace = */ true,
+      /*assertMode = */ false, /*recordLineSeparator = */ true
+    )
+    scanner.setSource(source.chars)
+
+    val buffer = new ArrayBuffer[Token]()
+
+    def loop(token: Int): Unit = {
+      if (token != ITerminalSymbols.TokenNameEOF) {
+        mapToken(token, scanner, source, buffer)
+        loop(scanner.getNextToken())
+      }
+    }
+
+    loop(scanner.getNextToken())
+    val array = buffer.toArray
+    Tokens(array, 0, array.size)
+  }
+
+  private def mapToken(
+      i: Int,
+      scanner: IScanner,
+      source: Input,
+      buffer: ArrayBuffer[Token]
+  ): Unit = {
+    val start = scanner.getCurrentTokenStartPosition()
+    val end = scanner.getCurrentTokenStartPosition()
+    def token(create: (Input, Dialect, Int, Int, String) => Token): Unit = {
+      buffer += create(
+        source,
+        dialects.Scala212,
+        start,
+        end,
+        scanner.getRawTokenSource().mkString
+      )
+    }
+    def keyword(create: (Input, Dialect, Int) => Token): Unit = {
+      buffer += create(source, dialects.Scala212, start)
+    }
+
+    def constant[T](
+        create: (Input, Dialect, Int, Int, T) => Token,
+        translate: String => T
+    ): Unit = {
+      buffer += create(
+        source,
+        dialects.Scala212,
+        start,
+        end,
+        translate(scanner.getRawTokenSource().mkString)
+      )
+    }
+
+    i match {
+      case ITerminalSymbols.TokenNameCOMMENT_BLOCK => token(Token.Comment.apply)
+      case ITerminalSymbols.TokenNameWHITESPACE =>
+        scanner.getCurrentTokenSource().foreach {
+          case '\n' => keyword(Token.LF.apply)
+          case '\r' => keyword(Token.CR.apply)
+          case '\f' => keyword(Token.FF.apply)
+          case '\t' => keyword(Token.Tab.apply)
+          case ' ' => keyword(Token.Space.apply)
+          case _ =>
```

Comment:
Can `ITerminalSymbols` be imported to shorten names? Also, what do you think about `// format: off` before pattern match? I find it almost unreadable if code blocks after `=>` are sometimes at the same line, sometimes at the newline. Having all of them at the same line as `=>` looks far better.